gtsam/basis/Basis.h

```diff
@@ -0,0 +1,373 @@
+ */
+
+namespace gtsam {
+
+/// CRTP Base class for function bases
+template <typename DERIVED>
+class Basis {
+ public:
+  /// Call weights for all x in vector, returns M*N matrix where M is the size
+  /// of the vector X.
+  static Matrix WeightMatrix(size_t N, const Vector& X) {
+    Matrix W(X.size(), N);
+    for (int i = 0; i < X.size(); i++)
+      W.row(i) = DERIVED::CalculateWeights(N, X(i));
+    return W;
+  }
+
+  /// Call weights for all x in vector, with interval [a,b]
+  /// Returns M*N matrix where M is the size of the vector X.
+  static Matrix WeightMatrix(size_t N, const Vector& X, double a, double b) {
+    Matrix W(X.size(), N);
+    for (int i = 0; i < X.size(); i++)
+      W.row(i) = DERIVED::CalculateWeights(N, X(i), a, b);
+    return W;
+  }
+
+  /**
+   * EvaluationFunctor at a given x, applied to Parameters.
+   * This functor is used to evaluate a parameterized function at a given scalar
+   * value x. When given a specific N*1 vector of Parameters, returns the scalar
+   * value of the function at x, possibly with Jacobian wrpt the parameters.
+   */
+  class EvaluationFunctor {
+   protected:
+    typename DERIVED::Weights weights_;
+
+   public:
+    // Used by FunctorizedFactor
+    using argument_type = Vector;
+    using return_type = double;
+
+    /// Constructor with interval [a,b]
+    EvaluationFunctor(size_t N, double x)
+        : weights_(DERIVED::CalculateWeights(N, x)) {}
+
+    /// Constructor with interval [a,b]
+    EvaluationFunctor(size_t N, double x, double a, double b)
+        : weights_(DERIVED::CalculateWeights(N, x, a, b)) {}
+
+    /// Regular 1D evaluation
+    double apply(const typename DERIVED::Parameters& f,
+                 OptionalJacobian<-1, -1> H = boost::none) const {
+      if (H) *H = weights_;
+      return (weights_ * f)(0);
+    }
+
+    /// c++ sugar
+    double operator()(const typename DERIVED::Parameters& f,
+                      OptionalJacobian<-1, -1> H = boost::none) const {
+      return apply(f, H);  // might call apply in derived
+    }
+
+    void print(const std::string& s = "") const {
+      std::cout << s << (s != "" ? " " : "") << weights_ << std::endl;
+    }
+  };
+
+  /**
+   * VectorEvaluationFunctor at a given x, applied to MatrixMN.
+   * This functor is used to evaluate a parameterized function at a given scalar
+   * value x. When given a specific M*N parameters, returns an M-vector the M
+   * corresponding functions at x, possibly with Jacobian wrpt the parameters.
+   */
+  template <int M>
+  class VectorEvaluationFunctor : protected EvaluationFunctor {
+   protected:
+    typedef Eigen::Matrix<double, M, 1> VectorM;
+    typedef Eigen::Matrix<double, M, -1> MatrixMN;
+    typedef Eigen::Matrix<double, /*MxMN*/ M, -1> Jacobian;
+    Jacobian H_;
+    void calculateJacobian() {
+      typedef Eigen::Matrix<double, M, M> MatrixM;
+      H_ = Eigen::kroneckerProduct(this->weights_, MatrixM::Identity());
```

Comment:
The [documentation](https://eigen.tuxfamily.org/dox/unsupported/group__KroneckerProduct__Module.html) warns against this syntax - it should be `kroneckerProduct(...).eval()`

Reply:
That's only if we are replacing a matrix, which is not the case here.

> If you want to replace a matrix by its Kronecker product with some matrix, do NOT do this: